pkg/util/http/client.go

```diff
@@ -31,86 +27,68 @@ func SanitizeURL(message string) string {
 	return apiKeyRegExp.ReplaceAllString(message, apiKeyReplacement)
 }
 
-// CreateHTTPTransport creates an *http.Transport for use in the agent
-func CreateHTTPTransport() *http.Transport {
-	tlsConfig := &tls.Config{
-		InsecureSkipVerify: config.Datadog.GetBool("skip_ssl_validation"),
+// Client wraps (http.Client).Do and resets the underlying connections at the
+// configured interval
+type Client struct {
+	httpClient    *http.Client
+	newHTTPClient func() *http.Client
+	resetInterval time.Duration
+
+	lastReset time.Time
+	mutex     sync.RWMutex
+}
+
+// NewClient returns an initialized Client resetting connections at the passed resetInterval.
+// The underlying http.Client used will be created using the passed http client factory.
+func NewClient(resetInterval time.Duration, newHTTPClient func() *http.Client) *Client {
+	return &Client{
+		resetInterval: resetInterval,
+		httpClient:    newHTTPClient(),
+		newHTTPClient: newHTTPClient,
+	}
+}
+
+// Do wraps (http.Client).Do. Thread safe.
+func (c *Client) Do(req *http.Request) (*http.Response, error) {
+	c.mutex.RLock()
+	httpClient := c.httpClient
+	c.mutex.RUnlock()
+	if c.shouldReset() {
+		log.Debug("Resetting HTTP client's connections")
+		c.resetHTTPClient()
 	}
 
-	if config.Datadog.GetBool("force_tls_12") {
-		tlsConfig.MinVersion = tls.VersionTLS12
+	return httpClient.Do(req)
+}
+
+// shouldReset returns whether the http.Client should be reset
+func (c *Client) shouldReset() bool {
```

Comment:
It might be simpler to just have one function `c.checkReset` which both checks if it should and does it. It would remove the `if` in `Do(..)`, the `return false` statements would just become `return` and the `return true` can be replaced by the two lines in `resetHTTPClient`. It would also acquire lock just once.

The fastest way would probably be the above, but only read-locking most of the time:

```go
c.mu.RLock()
sinceReset := time.Since(c.lastReset)
resetInterval := c.resetInterval
c.mu.RUnlock()
if sinceReset < resetInterval {
    return
}
c.mu.RLock()
// ...
```

Reply:
addressed, thanks. I kept a larger-scoped lock so that `c.lastReset` is consistently checked (and updated if necessary), even if multiple goroutines call `checkReset` concurrently